patroni/ha.py

```diff
@@ -1938,23 +1933,28 @@ def get_remote_member(self, member: Union[Leader, Member, None] = None) -> Remot
         name = member.name if member else 'remote_member:{}'.format(uuid.uuid1())
         return RemoteMember(name, data)
 
-    def get_failover_candidates(self, check_sync: bool = False) -> List[Member]:
-        """Return list of candidates for either manual or automatic failover.
+    def get_failover_candidates(self, check_synchronous: bool, exclude_candidate: bool) -> List[Member]:
+        """Return a list of candidates for either manual or automatic failover.
 
-        Mainly used to later be passed to ``Ha.is_failover_possible()``.
+        Exclude non-sync members when in synchronous mode, the current node (its checks are always performed earlier)
+        and the candidate if required. If failover candidate exclusion is not requested and a candidate is specified
+        in the /failover key, return the candidate only.
+        The result is further evaluated in the caller :func:`Ha.is_failover_possible` to check if any member is actually
+        healthy enough and is allowed to poromote.
 
-        :param check_sync: if ``True``, also check against the sync key members
+        :param check_synchronous: if ``True``, also check against the sync key members.
+        :param exclude_candidate: if ``True``, exclude :attr:`failover.candidate` from the candidates.
 
-        :returns: a list of ``Member`` ojects or an empty list if there is no candidate available
+        :returns: a list of :class:`Member` ojects or an empty list if there is no candidate available.
         """
-        failover = self.cluster.failover
-        if check_sync:
+        def is_eligible(node_name: str, check_synchronous: bool, exclude_candidate: bool) -> bool:
             # TODO: allow manual failover (=no leader specified) to async node
-            # every sync_standby or the candidate specified if is in sync_standbys
-            return [m for m in self.cluster.members
-                    if self.cluster.sync.matches(m.name)
-                    and (not failover or not failover.candidate or m.name == failover.candidate)]
-        else:
-            # every member or the candidate specified
-            return [m for m in self.cluster.members
-                    if not failover or not failover.candidate or m.name == failover.candidate]
+            failover = self.cluster.failover
+            exclude = [self.state_handler.name] + ([failover.candidate] if failover and exclude_candidate else [])
+            if check_synchronous and self.is_synchronous_mode() and not self.cluster.sync.is_empty and \
+               not self.cluster.sync.matches(node_name):
+                return False
+            return node_name not in exclude and \
```

Comment:
The following will help to get rid of https://github.com/zalando/patroni/pull/2804/files#diff-b6a3a0ef64783a6d32feb908d74a3a009e9f25008b8ca5313c9c9747e3b74a63R925
```suggestion
            return node.name not in exclude and and not node.nofailover and node.api_url \
```

Reply:
with this I feel like we start to mix up "eligibility" and "healthiness"
maybe we need to define the terms then